static/app/utils/useLocalStorageState.ts

```diff
@@ -0,0 +1,157 @@
+  }
+  if (value instanceof RegExp) {
+    throw new TypeError(makeTypeExceptionString('RegExp'));
+  }
+  if (SUPPORTS_MAP && value instanceof Map) {
+    throw new TypeError(makeTypeExceptionString('Map'));
+  }
+  if (SUPPORTS_SET && value instanceof Set) {
+    throw new TypeError(makeTypeExceptionString('Set'));
+  }
+  if (SUPPORTS_WEAK_MAP && value instanceof WeakMap) {
+    throw new TypeError(makeTypeExceptionString('WeakMap'));
+  }
+  if (SUPPORTS_WEAK_SET && value instanceof WeakSet) {
+    throw new TypeError(makeTypeExceptionString('WeakSet'));
+  }
+  return value;
+}
+
+function stringifyForStorage(value: unknown) {
+  return JSON.stringify(value, strictReplacer, 0);
+}
+
+function defaultOrInitializer<S>(
+  defaultValueOrInitializeFn: S | ((value?: unknown, rawStorageValue?: unknown) => S),
+  value?: unknown,
+  rawValue?: unknown
+): S {
+  if (typeof defaultValueOrInitializeFn === 'function') {
+    // https://github.com/microsoft/TypeScript/issues/37663#issuecomment-759728342
+    // @ts-expect-error
+    return defaultValueOrInitializeFn(value, rawValue);
+  }
+  return defaultValueOrInitializeFn;
+}
+
+// Initialize state with default value or value from localStorage.
+// If window is not defined uses the default value and **does not** throw an error
+function initializeStorage<S>(
+  key: string,
+  defaultValueOrInitializeFn: S | ((rawStorageValue?: unknown) => S)
+): S {
+  if (typeof key !== 'string') {
+    throw new TypeError('useLocalStorage: key must be a string');
+  }
+
+  // Return default if env does not support localStorage. Passing null to initializer
+  // to mimick not having any previously stored value there.
+  if (!SUPPORTS_LOCAL_STORAGE) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, null);
+  }
+
+  // getItem and try and decode it, if null is returned use default initializer
+  const jsonEncodedValue = localStorage.getItem(key);
+  if (jsonEncodedValue === null) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, null);
+  }
+
+  // We may have failed to parse the value, so just pass it down raw to the initializer
+  const decodedValue = tryParseStorage<S>(jsonEncodedValue);
+  if (decodedValue === null) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, jsonEncodedValue);
+  }
+
+  // We managed to decode the value, so use it
+  return defaultOrInitializer(defaultValueOrInitializeFn, decodedValue, jsonEncodedValue);
+}
+
+// Mimicks the behavior of React.useState but keeps state synced with localStorage.
+// The only difference from React is that when a state initializer fn is passed,
+// the first argument to that function will be the value that we decoded from localStorage
+// and the second argument will be the raw value from localStorage. This is useful for cases where you may
+// want to recover the error, apply a transformation or use an alternative parsing function.
+export function useLocalStorageState<S>(
```

Comment:
Very thorough implementation!

One extra feature you definitely don't need to add but I've found useful for useLocalStorage hooks is notifying other uses of the hook when updates are made. So you're able to use it in multiple locations and share the same state. Like in this implementation: https://usehooks-ts.com/react-hook/use-local-storage

(and once again just a discussion point, you definitely don't need this feature and may 
have a good reason for not wanting it)

Reply:
I actually thought about this and I had it pretty much done, but ended up removing it. 

I decided to remove it because I do not require it and because it feels like it should be simple enough to add later. You could always build a separate `useSynchonizedLocalStorageState` using the `useLocalStorageHook`, in which case I'd prefer to let this hook just take care of the syncing with storage.